File: binary-tree-leval-order-traversal-ii.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<ArrayList<Integer>> levelOrderBottom(TreeNode root) {
        ArrayList<ArrayList<TreeNode>> queue =
            new ArrayList<ArrayList<TreeNode>>();
        ArrayList<ArrayList<Integer>> ans = new ArrayList<ArrayList<Integer>>();
        if (root == null) return ans;
        queue.add(new ArrayList<TreeNode>());
        queue.get(0).add(root);
        while (true) {
            ArrayList<TreeNode> lastLevel = queue.get(queue.size() - 1);
            ArrayList<TreeNode> newLevel = new ArrayList<TreeNode>();
            for (TreeNode node : lastLevel) {
                if (node.left != null) newLevel.add(node.left);
                if (node.right != null) newLevel.add(node.right);
            }
            if (newLevel.isEmpty()) break;
            queue.add(newLevel);
        }
        for (ArrayList<TreeNode> nodes : queue) {
            ArrayList<Integer> values = new ArrayList<Integer>();
            for (TreeNode node :nodes) values.add(node.val);
            ans.add(values);
        }
        Collections.reverse(ans);
        return ans;
    }
}

File: binary-tree-leval-order-traversal-ii.cc
#include <cstdlib>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  vector<vector<int> > levelOrderBottom(TreeNode* root) {
    vector<vector<int> > ans;
    if (!root) return ans;
    vector<int> num_level;
    vector<TreeNode*> order;
    num_level.push_back(1);
    order.push_back(root);
    int next_node = 0;
    while (true) {
      int num_next_level = 0;
      for (int i = 0; i < num_level.back(); i++) {
        TreeNode* node = order[next_node++];
        if (node->left) {
          num_next_level++;
          order.push_back(node->left);
        }
        if (node->right) {
          num_next_level++;
          order.push_back(node->right);
        }
      }
      if (num_next_level > 0) {
        num_level.push_back(num_next_level);
      } else {
        break;
      }
    }
    int last = order.size() - 1;
    for (int i = num_level.size() - 1; i >= 0; i--) {
      vector<int> ids;
      for (int j = last - num_level[i] + 1; j <= last; j++) {
        ids.push_back(order[j]->val);
      }
      ans.push_back(ids);
      last -= num_level[i];
    }
    return ans;
  }
};
